### PR TITLE
WIP: use dynamic import to load es module tests.

### PR DESCRIPTION
**Pull request by harrysarson**
### Description of the Change

Uses the new (and as yet highly experimental) support in node for dynamically loading modules using `const obj = await import('./file-name.mjs');`.

I have added a new command line flag `--es-modules` which causes mocha to use `import()` instead of `require()` to load test files.

This allows tests to `import obj from './somefile.mjs'` without having to transpile tests before running them.

Obviously this only works in the very latest node version. This commit does not break mocha in `node <= 8` unless the user passes the `es-modules` flag.

Currently running a script in node with `--experimental-modules` fails if the file does not have an extension which means that `node --experimental-modules bin/_mocha` files. From my understanding (see https://github.com/nodejs/node/pull/18728) node v9.7 will be able to run scripts without the .js extension.

### Alternate Designs

Current design works fine as long as users are happy to use a transpiler.

### Why should this be in core?

Cannot be implemented outside of core.

### Benefits

Tests can be written as es modules.

### Possible Drawbacks

Current support for modules in node is experimental. It is likely that mocha will want to wait untill node starts properly supporting modules before merging this pr.

### Applicable issues

#3006 

Probably semver minor but could be major.

### CODE REVIEW

**Comment by boneskull** [comment]
This seems to break AppVeyor's build, but why?

File: .travis.yml [hunk 1]
```diff
@@ -14,17 +14,7 @@ matrix:
   fast_finish: true
   include:
     - node_js: '9'
-      env: TARGET=test.node COVERAGE=true
-    - node_js: '8'
```
Comment (boneskull) [inline .travis.yml]: we can't just remove all of these from the build matrix...

File: bin/mocha [hunk 2]
```diff
@@ -10,7 +10,7 @@
 const spawn = require('child_process').spawn;
 const path = require('path');
 const getOptions = require('./options');
-const args = [path.join(__dirname, '_mocha')];
+const args = [path.join(__dirname, '_mocha.js')];
```
Comment (boneskull) [inline bin/mocha]: why is this necessary?

File: lib/mocha.js [hunk 3]
```diff
@@ -234,6 +234,34 @@ Mocha.prototype.loadFiles = function (fn) {
   fn && fn();
 };
 
+/**
+ * Load registered files via es6 dynamic imports.
+ *
+ * Note: `eval` is used in this function or else mocha will not
+ * compile on older node versions.
```
Comment (boneskull) [inline lib/mocha.js]: I don't understand where this is supposed to work.

File: lib/mocha.js [hunk 4]
```diff
@@ -234,6 +234,34 @@ Mocha.prototype.loadFiles = function (fn) {
...
+ */
+Mocha.prototype.dynamicallyImportFiles = function (fn) {
+  var self = this;
+  var suite = this.suite;
+  Promise
+    .all(this.files.map(function (file) {
+      file = path.resolve(file);
+      suite.emit('pre-require', global, file, self);
+      return eval('import(file)')
+        .then(function (module) {
```
Comment (boneskull) [inline lib/mocha.js]: probably shouldn't use `module`

File: lib/mocha.js (hunk 3 above)
Comment (harrysarson) [reply lib/mocha.js]: The new `import()` operator allows ECMAScript Modules to be loaded dynamically. The hope was that tests written as modules (i.e. tests including `import obj from './script.js'` statements) can be loaded into Mocha using dynamic imports without having to first use a transpiler to convert ECMAScript Modules into common js modules.

Comment (boneskull) [reply lib/mocha.js]: right, but I'm unclear on why we're removing old versions of Node.js from the matrix.

> Current support for modules in node is experimental. It is likely that mocha will want to wait untill node starts properly supporting modules before merging this pr.

if we're going to break backwards compatibility, then it'd have to wait until the oldest version of Node.js that supports modules is no longer maintained.  that'll be a few years.  so, that's why I'm trying to understand what I'm supposed to do with this...

Comment (harrysarson) [reply lib/mocha.js]: Sorry, I didn't mean to add that change to the pull request, I will revert it.

The idea was that mocha could add a new CLI flag which would cause mocha to use the new `import()` instead of `require()` to load test files.

This new CLI flag could only be used on the most recent versions of node but mocha would work normally if this flag was not set.

The tests fail because I added a test for the new flag and travis/appveyor runs this test on old versions of node, I didn't know how to work around that.

**Comment by demurgos** [comment]
Why was this PR closed?

I am currently using Mocha together with `@std/esm` to test my .mjs files but this is suboptimal (edge cases around interop with cjs or code coverage). Now that dynamic imports are available it would be nice to be able to use ESM directly with Mocha.

**Comment by boneskull** [comment]
Until modules are unflagged and/or no longer considered "experimental", I'm concerned about adopting something too soon.  Specifically, when there's agreement in core around interop between `require()` and `import`, filename extensions, etc.--then it's probably safe to move.  Too early, and it'll be expensive to fix any mistakes we made by jumping on it too soon.

**Comment by demurgos** [comment]
Thanks for the reply: it's a fair reason.

I was worried that there was an issue with the PR preventing it from working. Still, it's good to know that it's a Node stability issue instead and that you are able to add support for it when possible.

**Comment by harrysarson** [comment]
I was a bit optimistic with this PR, definately need to wait until modules are finalised in node.

The other problem, which is why my work was so hacky, is that the dynamic `import()` loads modules asynchronously whereas mocha currently loads tests synchronously (using `require()`) so it is not as simple as just switching one for the other.

**Comment by demurgos** [comment]
Hi,
The modules discussion is still ongoing, but it seems that there's agreement that the imports will be async.

A possible way to prepare the support for ESM would be to force the current imports to be async by default: replace the resolution code from `require("test.spec.js")` to `Promise.resolve(require("test.spec.js"))`.
Replacing the inner `require` by `import` would then be trivial.

Looking at the PR, it seems that most of the code is already async. I am not familiar enough with mocha's codebase so maybe my suggestion is not needed. I just hope that Mocha will be able to support ESM when available.

**Comment by harrysarson** [comment]
@demurgos from memory it should be fairly straightforward to make that change
